neps/nep-519-yield-execution.md

```diff
@@ -0,0 +1,164 @@
+---
+NEP: 519
+Title: Yield Execution
+Authors: Akhi Singhania <akhi3030@gmail.com>
+Status: Draft
+DiscussionsTo: https://github.com/near/NEPs/pull/519
+Type: Protocol
+Version: 0.0.0
+Created: 2023-11-17
+LastUpdated: 2023-11-20
+---
+
+## Summary
+
+Today, when a smart contract is called by a user or another contract, it has no sensible way to delay responding to the caller.  There exist use cases where contracts would benefit from being able to delay responding till some arbitrary time in the future.  This proposal introduces such possibility into the NEAR protocol.
+
+## Motivation
+
+There exist some situations where when a smart contract on NEAR is called, it will only be able to provide an answer at some arbitrary time in the future.  So the callee needs a way to defer replying to the caller till this time in future.
+
+Examples include when a smart contract (`S`) provides MPC signing capabilities parties external to the NEAR protocol are computing the signature.  The rough steps are:
+1. Signer contract provides a function `fn sign_payload(Payload, ...)`.
+2. When called, the contract updates some contract state which is being monitored by external indexers to indicate that a new signing request has been received.  It also defers replying to the caller.
+3. The indexers observe the new signing request, they compute a signature and call another function `fn signature_available(Signature, ...)` on the signer contract.
+4. The signer contract validates the signature and if validate, replies to the original caller.
+
+Today, the NEAR protocol has no sensible way to defer replying to the caller in step 2 above.  This proposal proposes adding two following new host functions to the NEAR protocol:
+
+- `yield`: this can be called by a contract to indicate to the protocol that it is not ready yet to reply to its caller.
+- `resume`: a contract can use this mechanism to indicate to a protocol that it is now ready to reply to a caller that it had deferred earlier.
+
+If these two host functions were available, then `yield` would be used in step 2 above and `resume` would be used in step 4 above.
```

Comment:
Can you expand on the motivation section why is the new interface of `yield` / `resume` essential for this workflow. What you describe can be done with two transactions today. 

1. Signer contract provides a function `fn sign_payload(Payload, ...)` **[SAME]**
2. When called, the contract updates some contract state which is being monitored by external indexers to indicate that a new signing request has been received. It also defers replying to the caller **on a future transaction**.
3. The indexers observe the new signing request, they compute a signature and call another function **in a new transaction** `fn signature_available(Signature, ...)` on the signer contract. 
4. The signer contract validates the signature and if validate, replies to the original caller. **[SAME]**

This workflow requires two transactions (on step 1 by the user, and on step 3 by the MPC network).

Reply:
>  It also defers replying to the caller on a future transaction.

How can a contract do that today?  My understanding is that if contract A calls B today, one of the following things must happen:

1. B must provide a reply
2. B calls another contract (potentially itself)
3. If B does not provide a reply or traps or runs out of gas, etc. then the protocol generates a response on its behalf.

Are you proposing option 2 above?

Reply:
I think Marcelo is proposing option 1 and there is a contract-level protocol for these kinds of off-chain computation. Consider the following sketch implementation:

```rust
struct FunctionCall {
    contract: AccountId,
    method_name: String,
    gas: u64,
    deposit: u128,
}

struct Request<T> {
    input: T,
    callback: FunctionCall,
}

enum Response<T> {
    // `id` computed as the hash of the `Request` concatenated with any relevant environment data (e.g. timestamp).
    // We will assume that requests with identical `id` values will also have identical results when the computation completes.
    Pending { id: CryptoHash },
    Complete { value: T },
    // An error case to allow the contract to reject computations it doesn't want to do.
    InvalidRequest { message: String },
}

fn off_chain_computation(&mut self, request: Request) -> Response {
    let id = hash_with_environment(&request);
    // Caching could be optional
    if let Some(value) = self.cache.get(id) {
        return Response::Complete { value };
    }
    // Verify that env::deposit() >= request.callback.deposit  + storage_cost(request),
    // along with any other business logic that might determine an invalid request.
    if let Err(message) = self.validate_request(&request) {
        return Response::InvalidRequest { message};
    }
    self.requests.insert(id, request);
    // Maybe also emit an event here to make indexer's lives easier
    Response::Pending
}

fn complete_computation(&mut self, id: CryptoHash, value: T) {
    // TODO: access control (what accounts can submit a response?)

    // Optional caching. Could make this some kind of LRU cache to bound storage usage.
    self.cache.insert(id, value);

    if let Some(request) = self.requests.remove(id) {
        // Maybe also emit an event here to make indexer's lives easier
        let cb = request.callback;
        // TODO: could consider refunding the storage cost of the request also since it has been removed.
        env::promise_return(cb.contract.call(cb.method_name).args(id, value).gas(cb.gas).deposit(cb.deposit));
    }
}
```

With the above scheme any contract calling `off_chain_computation` would need to be able to properly handle the `Response::Pending` case. Usually this will mean saving the `id` into its state so that when the callback is invoked later the contract will be able to resume whatever computation it wanted to do.

This scheme does not have an explicit timeout, it would be up to the caller to enforce that. For example it could save a timestamp with the request id and in the callback function check if there was a timeout. In the case the off-chain computation never completes then it would also be up to the caller to trigger their callback method themselves to do the on-chain timeout handling logic.

An important note with this scheme is that whatever service submits the result of the off-chain computation also pays for the gas needed to execute the callback. A service may limit the amount it is willing to pay by adding a request validation rule limiting the amount of gas attached to the callback. If the service severely limits the amount of gas it is willing to spend then it could be that the callback could only persist the response and the caller would need to submit another transaction that does further computation.

In my opinion, it is that "who pays" complexity which makes a scheme like this not a universal solution to this problem and does make a protocol change worthwhile.

Reply:
To make sure I understand this correctly, I think the flow is as following:

- Contract B provides some service that requires offchain computation.
- Contract A calls B to use the service
- B updates its state and then replies to A that the computation is in progress
- Some time later, the offchain computation finishes and B is called by the offchain service
- B sends another message to A providing the results of the offchain computation.

Is that correct?

Reply:
Option 2 you provided @akhi3030 is potentially interesting. The contract would call itself in a loop waiting to either run out of gas or see the response in its storage and send it back to the caller. The issue I see with that kind of scheme is how expensive it would be in terms of gas usage. I think it takes around 2 Tgas to make a cross-contract call (even a self-call), so with a 300 Tgas limit there would only be 150 blocks of wait time. And of course you could not dedicate the whole transaction to waiting for the response if you wanted to do any computation before or after, so maybe in practice you could only wait for at most 50 blocks (100 Tgas) to get a response and I am not sure if that is long enough.

Reply:
> To make sure I understand this correctly, I think the flow is as following:
> 
> * Contract B provides some service that requires offchain computation.
> * Contract A calls B to use the service
> * B updates its state and then replies to A that the computation is in progress
> * Some time later, the offchain computation finishes and B is called by the offchain service
> * B sends another message to A providing the results of the offchain computation.
> 
> Is that correct?

Yes, that is correct

Reply:
I think @DavidM-D 's team is currently using option 2.  They view it as a hacky workaround till a more proper solution is available.

Option 1 above may not be generally possible as it does not follow the usual request - response procedure.  As contract A (in the above example) is not always under our control.  

The way I see this, we in protocol need requirements for what kind of usecase needs to be supported.  If the current protocol is good enough, that is great and nothing to be done 😃

Reply:
Yup, we're not really happy with the call loop approach, it's tremendously brittle and will fail if you do simple things, like call for multiple signatures.

To some extent you're correct that we can pick any arbitrary API so we can use submit and poll systems. The problem is that every contract that calls this endpoint will also have to have application level submit and poll. It also means that we need to design a version of the indexer that is call stack aware. This would be a tremendous amount of work and would lead to function colouring for anything that has a signature anywhere in it's call tree.

Reply:
> Yup, we're not really happy with the call loop approach, it's tremendously brittle and will fail if you do simple things, like call for multiple signatures.

Isn't the idea elaborated by Michael (https://github.com/near/NEPs/pull/519#discussion_r1444876559) a better and more robust approach than a loop? 
- Is it a concern who pays for the gas in the second transaction (maybe it could be solved with delegated transactions).
- Or is it a concern about the engineering complexity of this approach? I found it feasible but complex.

Btw what Michel described is how Aurora works internally, so there is some experience in the ecosystem with those patterns, and we know they work.

> we can pick any arbitrary API so we can use submit and poll systems. The problem is... 

You described a lot of complexity on the application layer without this feature, but notice that the alternative that is being proposed here is to move this complexity into the protocol, which IMO is a bad idea. The protocol should be as simple as possible (it is not simple by any means today, but we should try to avoid further complexity if possible).

This change is quite invasive, and it is hard to foresee unintended consequences. 
- This requires updating the SDK
- Documentation to ensure developers know how to use this feature properly and don't shoot themselves in the foot.
- Increase the bar to hire auditors, given they need to be aware of this quite specific feature in NEAR-Protocol
- Hard to maintain at the protocol level
- Any future change to the protocol, including gas/balance accounting, must make sure yield/resume keeps working (backward compatibility)
- This feature is quite broad in what it enables, but it is being designed with one application in mind.

Reply:
Based on our last discussion, I believe it is fine to resolve this now.